### PR TITLE
Bad operations

### PR DESCRIPTION
**Pull request by JohnHamby**
This change set adds an IsInvalid property to IOperation, and adds InvalidExpression and InvalidStatement to OperationKind. These make it easier for operation actions to be robust in analyzing semantically malformed code.

@srivatsn @mavasani @heejaechang @genlu @nguerrera @AlekseyTs @CyrusNajmabadi might take an interest.

Addresses #7221

### CODE REVIEW

File: src/Compilers/CSharp/Portable/BoundTree/Statement.cs [hunk 1]
```diff
@@ -9,6 +9,8 @@ partial class BoundStatement : IStatement
     {
         OperationKind IOperation.Kind => this.StatementKind;
 
+        bool IOperation.HasErrors => this.HasErrors;
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/BoundTree/Statement.cs]: Why do we need this?  This seems like an extra unnecessary indirectoin.  Why not just have the existing HasErrors implement IOperation.HasErrors instead?

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs [hunk 2]
```diff
@@ -594,7 +596,7 @@ partial class BoundIncrementOperator : IIncrementExpression
 
     partial class BoundBadExpression
     {
-        protected override OperationKind ExpressionKind => OperationKind.None;
+        protected override OperationKind ExpressionKind => OperationKind.BadExpression;
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: "Bad" seems like a really "bad" name.  Why not "Invalid" or something lke that.  I

File: src/Compilers/Core/Portable/Compilation/Expression.cs [hunk 3]
```diff
@@ -104,6 +104,8 @@ public ConditionalChoice(IExpression condition, IExpression ifTrue, IExpression
 
         public OperationKind Kind => OperationKind.ConditionalChoiceExpression;
 
+        public bool HasErrors => false;
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Compilation/Expression.cs]: Could we not have a common base type that these all get this functionality from?

File: src/Compilers/Core/Portable/Compilation/IOperation.cs [hunk 4]
```diff
@@ -14,6 +14,12 @@ public interface IOperation
         /// Identifies the kind of the operation.
         /// </summary>
         OperationKind Kind { get; }
+
+        /// <summary>
+        /// Indicates whether the compiler has reported semantic errors about the operation.
+        /// </summary>
+        bool HasErrors { get; }
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Compilation/IOperation.cs]: Can you be more specific?  Does this mean an error directly at this operation level, or an error anywhere underneath? 

And, if this is the case, can we possibly have the errors also surfaced some way?

Also, if htis is only about Semantic errors, then maybe "HasSemanticErrors" would be better.  I would have had no idea this didn't include syntax based on hte name.

File: src/Compilers/Core/Portable/Compilation/IOperation.cs [hunk 5]
```diff
@@ -27,6 +33,7 @@ public enum OperationKind
     {
         None,
 
+        BadStatement,
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Compilation/IOperation.cs]: Same comment about "Bad"

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs (hunk 2 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: The usual feedback has been to surface names used in the bound trees, at least when those names are common across languages. I've used the term "Bad" for this sort of thing for so long that for me it seems natural, but I'm happy to use another term.

"Invalid" is rejected by some policy checkers which recommend "disabled" or some such instead.

File: src/Compilers/CSharp/Portable/BoundTree/Statement.cs (hunk 1 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/BoundTree/Statement.cs]: Personally, I like seeing the entire interface implementation, but of course it's not necessary. Except for giving a method inliner some exercise, it doesn't cost anything either.

File: src/Compilers/Core/Portable/Compilation/IOperation.cs (hunk 4 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Compilation/IOperation.cs]: The indicated errors can be either syntactic or semantic. The intent is to indicate that the individual operation is semantically unreliable. "DoNotTrustMe", "SomethingIsWrongWithMe", or "IsBad" would be accurate names, but I can almost guarantee that if I change the name someone else will want it brought into line with the name in the bound trees.

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs (hunk 2 above)
Comment (davkean) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: "Invalid" is an extremely common term in .NET, for example, InvalidOperationException

Comment (nguerrera) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: I agree that 'Invalid'  would be fine and that would be what I would have chosen. (We must have turned off any policy against it long ago in the framework)

FWIW, we have also have at least one precedent for 'Bad' in .NET: BadImageFormatException. :)

Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: OK, how about "InvalidExpression" and "InvalidStatement" operation kinds, and "IsInvalid" for the new property?

File: src/Compilers/VisualBasic/Portable/BoundTree/Statement.vb [hunk 6]
```diff
@@ -582,6 +594,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             End Get
         End Property
 
+        Private ReadOnly Property IHasErrors As Boolean Implements IOperation.HasErrors
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/BoundTree/Statement.vb]: Is this property not implemented by a base class?

**Comment by AlekseyTs** [comment]
LGTM

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: That's what I would prefer.

File: src/Compilers/CSharp/Portable/BoundTree/Statement.cs (hunk 1 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/BoundTree/Statement.cs]: A more correct implementation is coming in the next iteration, which has different behavior for each class.

File: src/Compilers/Core/Portable/Compilation/Expression.cs (hunk 3 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Compilation/Expression.cs]: A more correct implementation actually has different behavior per type. This is coming in the next iteration.

File: src/Compilers/VisualBasic/Portable/BoundTree/Statement.vb (hunk 6 above)
Comment (JohnHamby) [inline src/Compilers/VisualBasic/Portable/BoundTree/Statement.vb]: It is now, but won't be after renaming the property to "IsInvalid".

**Comment by JohnHamby** [comment]
I've pushed a new iteration.

**Comment by genlu** [comment]
:+1:

File: src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb [hunk 7]
```diff
@@ -61,6 +61,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             End Get
         End Property
 
+        Private ReadOnly Property IIsInvalid As Boolean Implements IOperation.IsInvalid
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: "IsInvalid"?

File: src/Compilers/VisualBasic/Portable/BoundTree/Statement.vb [hunk 8]
```diff
@@ -15,6 +15,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             End Get
         End Property
 
+        Private ReadOnly Property IIsInvalid As Boolean Implements IOperation.IsInvalid
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/BoundTree/Statement.vb]: "IsInvalid"?

File: src/Compilers/VisualBasic/Portable/BoundTree/Statement.vb [hunk 9]
```diff
@@ -582,6 +594,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/BoundTree/Statement.vb]: "IsInvalid"?

File: src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb (hunk 7 above)
Comment (JohnHamby) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: This case follows a naming convention for implementations of IOperation members, prefacing the implemented member name with "I". One might question the appropriateness of the convention, but addressing that would be a separate change set.

File: src/Compilers/VisualBasic/Portable/BoundTree/Statement.vb (hunk 8 above)
Comment (JohnHamby) [inline src/Compilers/VisualBasic/Portable/BoundTree/Statement.vb]: Again, this case is following a convention.

File: src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb (hunk 7 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: Usually we prefix the name with the type name, i.e. ```IOperation_IsInvalid```

Comment (JohnHamby) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: OK, that would be better. I'll make that change systematically in a near-future PR.

**Pull request merged**